src/utils/fetch-with-cache.ts

```diff
@@ -0,0 +1,51 @@
+export interface CachedResponsePair {
+  etag: string;
+  response: Response;
+}
+
+type FetchOptions = Omit<RequestInit, 'headers'> & { headers?: Headers };
+
+export const cache = new Map<string, CachedResponsePair>();
+
+export async function fetchWithCache(
+  url: string,
+  options: FetchOptions = {},
+): Promise<Response> {
+  const { headers: customHeaders = new Headers(), ...customOptions } = options;
+  const authHeader = customHeaders.get('Authorization');
+  const cacheKey = `${url},${authHeader || ''}`;
+
+  const cachedResponsePair = cache.get(cacheKey);
+
+  if (cachedResponsePair) {
+    const { etag, response: cachedResponse } = cachedResponsePair;
+    const headers = new Headers(customHeaders);
+    headers.set('If-None-Match', etag);
+
+    const res = await fetch(url, { ...customOptions, headers });
+
+    if (res.status === 304) {
+      return cachedResponse.clone();
```

Comment:
the current implementation sort of works, but it always responds with the full previous response

a neater approach would likely be to mentally see `fetchWithCache` not as a more powerful fetch, but rather as a layer on top of fetch which only exposes the response headers, status code and data (not body)